src/js/node/net.js

```diff
@@ -408,32 +423,94 @@ const Socket = (function (InternalSocket) {
         this._securePending = true;
         if (connectListener) this.on("secureConnect", connectListener);
       } else if (connectListener) this.on("connect", connectListener);
-      bunConnect(
-        path
-          ? {
+
+      // start using existing connection
+
+      if (connection) {
+        const socket = connection[bunSocketInternal];
+
+        if (socket) {
+          const result = socket.wrapTLS({
+            data: this,
+            tls,
+            socket: Socket.#Handlers,
+          });
+          if (result) {
+            const [raw, tls] = result;
+            // replace socket
+            connection[bunSocketInternal] = raw;
+            raw.timeout(raw.timeout);
+            raw.connecting = false;
+            // set new socket
+            this[bunSocketInternal] = tls;
+            tls.timeout(tls.timeout);
+            tls.connecting = true;
+            this[bunSocketInternal] = socket;
+            // start tls
+            tls.open();
+          } else {
+            this[bunSocketInternal] = null;
+            throw new Error("Invalid socket");
+          }
+        } else {
+          // wait to be connected
+          connection.once("connect", () => {
+            const socket = connection[bunSocketInternal];
+            if (!socket) return;
+
+            const result = socket.wrapTLS({
               data: this,
-              unix: path,
-              socket: Socket.#Handlers,
               tls,
-            }
-          : {
-              data: this,
-              hostname: host || "localhost",
-              port: port,
               socket: Socket.#Handlers,
-              tls,
-            },
-      );
+            });
+
+            if (result) {
+              const [raw, tls] = result;
+              // replace socket
+              connection[bunSocketInternal] = raw;
+              raw.timeout(raw.timeout);
+              raw.connecting = false;
+              // set new socket
+              this[bunSocketInternal] = tls;
+              tls.timeout(tls.timeout);
+              tls.connecting = true;
+              this[bunSocketInternal] = socket;
+              // start tls
+              tls.open();
+            } else {
+              this[bunSocketInternal] = null;
+              throw new Error("Invalid socket");
+            }
+          });
+        }
+      } else if (path) {
+        // start using unix socket
+        bunConnect({
```

Comment:
Does this promise need to be awaited?

Reply:
actually, we need to emit error fixed at https://github.com/oven-sh/bun/pull/3457/commits/d4b6924931674d58f368d0298879e49c06405996